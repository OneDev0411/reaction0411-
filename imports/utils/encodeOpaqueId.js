import { curry } from "ramda";

/**
 * @name encodeOpaqueId
 * @method
 * @memberof GraphQL/Transforms
 * @summary Transforms an internal ID to an opaque ID
 * @param {String} namespace The namespace of the ID
 * @param {String} id The ID to transform
 * @returns {String} An opaque ID
 */
const encodeOpaqueId = curry((namespace, id) => {
  if (typeof id !== "string" && typeof id !== "number") return id;
  const unencoded = `${namespace}:${id}`;
  return Buffer.from(unencoded).toString("base64");
});

export default encodeOpaqueId;
