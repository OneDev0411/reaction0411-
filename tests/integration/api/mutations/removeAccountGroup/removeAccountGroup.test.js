import encodeOpaqueId from "@reactioncommerce/api-utils/encodeOpaqueId.js";
import importAsString from "@reactioncommerce/api-utils/importAsString.js";
import Factory from "/tests/util/factory.js";
import TestApp from "/tests/util/TestApp.js";

const AddAccountToGroupMutation = importAsString("./AddAccountToGroupMutation.graphql");
const RemoveAccountGroupMutation = importAsString("./RemoveAccountGroupMutation.graphql");

jest.setTimeout(300000);

let addAccountToGroup;
let removeAccountGroup;
let customerGroup;
let mockAdminAccount;
let mockAdminAccountWithBadPermissions;
let mockCustomerAccount;
let shopId;
let shopOpaqueId;
let testApp;
let testGroup;

let guestGroup;
let ownerGroup;
let shopManagerGroup;

beforeAll(async () => {
  testApp = new TestApp();
  await testApp.start();

  shopId = await testApp.insertPrimaryShop();
  shopOpaqueId = encodeOpaqueId("reaction/shop", shopId);

  addAccountToGroup = testApp.mutate(AddAccountToGroupMutation);
  removeAccountGroup = testApp.mutate(RemoveAccountGroupMutation);

  // Create users
  mockAdminAccount = Factory.Account.makeOne({
    _id: "mockAdminAccount",
    roles: {
      [shopId]: ["owner", "admin", "ownerGroupPermission", "testGroupGroupPermission", "customerGroupPermission"]
    },
    shopId
  });

  mockAdminAccountWithBadPermissions = Factory.Account.makeOne({
    _id: "mockAdminAccountWithBadPermissions",
    roles: {
      [shopId]: ["admin"]
    },
    shopId
  });

  mockCustomerAccount = Factory.Account.makeOne({
    _id: "mockCustomerAccount",
    groups: [],
    shopId
  });

  // Create group data mocks
  customerGroup = Factory.Group.makeOne({
    _id: "customerGroup",
    createdBy: null,
    name: "customer",
    permissions: ["customer"],
    slug: "customer",
    shopId
  });

  guestGroup = Factory.Group.makeOne({
    _id: "guestGroup",
    createdBy: null,
    name: "guest",
    permissions: ["guest"],
    slug: "guest",
    shopId
  });

  ownerGroup = Factory.Group.makeOne({
    _id: "ownerGroup",
    createdBy: null,
    name: "owner",
    permissions: ["owner"],
    slug: "owner",
    shopId
  });

  shopManagerGroup = Factory.Group.makeOne({
    _id: "shopManagerGroup",
    createdBy: null,
    name: "shop manager",
    permissions: ["admin"],
    slug: "shop manager",
    shopId
  });

  testGroup = Factory.Group.makeOne({
    _id: "testGroup",
    createdBy: null,
    description: "a group for testing purposes",
    name: "test-int-group",
    permissions: ["test-perm-1", "test-perm-2"],
    slug: "test-int-group",
    shopId
  });
});

// There is no need to delete any test data from collections because
// testApp.stop() will drop the entire test database. Each integration
// test file gets its own test database.
afterAll(() => testApp.stop());

beforeEach(async () => {
  // Create users
  await testApp.createUserAndAccount(mockAdminAccount);
  await testApp.createUserAndAccount(mockAdminAccountWithBadPermissions);
  await testApp.createUserAndAccount(mockCustomerAccount);
  await testApp.setLoggedInUser(mockAdminAccount);

  // Create groups
  await testApp.collections.Groups.insertOne(customerGroup);
  await testApp.collections.Groups.insertOne(guestGroup);
  await testApp.collections.Groups.insertOne(ownerGroup);
  await testApp.collections.Groups.insertOne(shopManagerGroup);
  await testApp.collections.Groups.insertOne(testGroup);

  // Add customer account to the testGroup
  await addAccountToGroup({
    accountId: encodeOpaqueId("reaction/account", "mockCustomerAccount"),
    groupId: encodeOpaqueId("reaction/group", "testGroup")
  });

  await testApp.clearLoggedInUser();
});

afterEach(async () => {
  await testApp.collections.Groups.deleteMany({});
  await testApp.collections.Accounts.deleteMany({});
  await testApp.collections.users.deleteMany({});
});

test("a customer account should not be able to remove a group", async () => {
  await testApp.setLoggedInUser(mockCustomerAccount);

  try {
    await removeAccountGroup({
      input: {
        groupId: encodeOpaqueId("reaction/group", "testGroup"),
        shopId: shopOpaqueId
      }
    });
  } catch (errors) {
    expect(errors).toMatchSnapshot();
    return;
  }
});

test("an admin account should be able to remove a group", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  // Expect the Account to be moved to the customer group
  const beforeAccount = await testApp.context.collections.Accounts.findOne({ _id: "mockCustomerAccount" });
  expect(beforeAccount.groups).toEqual(["testGroup"]);

  let result;

  try {
    result = await removeAccountGroup({
      input: {
        groupId: encodeOpaqueId("reaction/group", "testGroup"),
        shopId: shopOpaqueId
      }
    });
  } catch (error) {
    expect(error).toBeUndefined();
    return;
  }

  expect(result.removeAccountGroup.group._id).toEqual(encodeOpaqueId("reaction/group", "testGroup"));

  // Ensure the group was deleted
  const deletedGroup = await testApp.context.collections.Groups.findOne({ _id: testGroup });
  expect(deletedGroup).toBeNull();

  // Expect the Account to be moved to the customer group
  const afterAccount = await testApp.context.collections.Accounts.findOne({ _id: "mockCustomerAccount" });
  expect(afterAccount.groups).toEqual(["customerGroup"]);
});

test("an admin account cannot delete default groups, 'owner', 'shop manager', 'guest' or 'customer'", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  try {
    await removeAccountGroup({
      input: {
        groupId: encodeOpaqueId("reaction/group", "shopManagerGroup"),
        shopId: shopOpaqueId
      }
    });
  } catch (errors) {
    expect(errors).toMatchSnapshot();
  }

  try {
    await removeAccountGroup({
      input: {
        groupId: encodeOpaqueId("reaction/group", "ownerGroup"),
        shopId: shopOpaqueId
      }
    });
  } catch (errors) {
    expect(errors).toMatchSnapshot();
  }

  try {
    await removeAccountGroup({
      input: {
        groupId: encodeOpaqueId("reaction/group", "guestGroup"),
        shopId: shopOpaqueId
      }
    });
  } catch (errors) {
    expect(errors).toMatchSnapshot();
  }

  try {
    await removeAccountGroup({
      input: {
        groupId: encodeOpaqueId("reaction/group", "customerGroup"),
        shopId: shopOpaqueId
      }
    });
  } catch (errors) {
    expect(errors).toMatchSnapshot();
  }
});

test("an admin account cannot remove a group unless there is a default customer group", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  // Remove the default customer group directly with mongo.
  // This simulates a user manually deleting or somehow otherwise altering the default customer group in the database
  await testApp.collections.Groups.deleteOne({ _id: "customerGroup" });

  // Expect the Account to be moved to the customer group
  const beforeAccount = await testApp.context.collections.Accounts.findOne({ _id: "mockCustomerAccount" });
  expect(beforeAccount.groups).toEqual(["testGroup"]);

  try {
    await removeAccountGroup({
      input: {
        groupId: encodeOpaqueId("reaction/group", "testGroup"),
        shopId: shopOpaqueId
      }
    });
  } catch (errors) {
    expect(errors).toMatchSnapshot();
    return;
  }
});
