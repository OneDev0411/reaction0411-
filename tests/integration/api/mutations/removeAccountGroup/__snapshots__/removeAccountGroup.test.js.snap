// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a customer account should not be able to remove a group 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "FORBIDDEN",
      "exception": Object {
        "details": Object {},
        "error": "access-denied",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "Access Denied",
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 3,
      },
    ],
    "message": "Access Denied",
    "path": Array [
      "removeAccountGroup",
      "group",
    ],
  },
]
`;

exports[`an admin account cannot delete default groups, 'owner', 'shop manager', 'guest' or 'customer' 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "details": Object {},
        "error": "access-denied",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "Cannot remove default group with ID shopManagerGroup.",
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 3,
      },
    ],
    "message": "Cannot remove default group with ID shopManagerGroup.",
    "path": Array [
      "removeAccountGroup",
      "group",
    ],
  },
]
`;

exports[`an admin account cannot delete default groups, 'owner', 'shop manager', 'guest' or 'customer' 2`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "details": Object {},
        "error": "access-denied",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "Cannot remove default group with ID ownerGroup.",
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 3,
      },
    ],
    "message": "Cannot remove default group with ID ownerGroup.",
    "path": Array [
      "removeAccountGroup",
      "group",
    ],
  },
]
`;

exports[`an admin account cannot delete default groups, 'owner', 'shop manager', 'guest' or 'customer' 3`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "details": Object {},
        "error": "access-denied",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "Cannot remove default group with ID guestGroup.",
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 3,
      },
    ],
    "message": "Cannot remove default group with ID guestGroup.",
    "path": Array [
      "removeAccountGroup",
      "group",
    ],
  },
]
`;

exports[`an admin account cannot delete default groups, 'owner', 'shop manager', 'guest' or 'customer' 4`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "details": Object {},
        "error": "access-denied",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "Cannot remove default group with ID customerGroup.",
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 3,
      },
    ],
    "message": "Cannot remove default group with ID customerGroup.",
    "path": Array [
      "removeAccountGroup",
      "group",
    ],
  },
]
`;

exports[`an admin account cannot remove a group unless there is a default customer group 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "details": Object {},
        "error": "server-error",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "Cannot remove group testGroup. Default \\"customer\\" group doesn't exist to move users to.",
      },
    },
    "locations": Array [
      Object {
        "column": 5,
        "line": 3,
      },
    ],
    "message": "Cannot remove group testGroup. Default \\"customer\\" group doesn't exist to move users to.",
    "path": Array [
      "removeAccountGroup",
      "group",
    ],
  },
]
`;
